src/Messages.jl

```diff
@@ -0,0 +1,555 @@
+
+Set header `value` for `key` if it is not already set.
+"""
+
+function defaultheader(m, v::Pair)
+    if header(m, first(v)) == ""
+        setheader(m, v)
+    end
+    return
+end
+
+
+"""
+    ischunked(::Message)
+
+Does the `Message` have a "Transfer-Encoding: chunked" header?
+"""
+
+ischunked(m) = hasheader(m, "Transfer-Encoding", "chunked")
+
+
+"""
+    appendheader(::Message, key => value)
+
+Append a header value to `message.headers`.
+
+If `key` is `""` the `value` is appended to the value of the previous header.
+
+If `key` is the same as the previous header, the `vale` is [appended to the
+value of the previous header with a comma
+delimiter](https://stackoverflow.com/a/24502264)
+
+`Set-Cookie` headers are not comma-combined because [cookies often contain
+internal commas](https://tools.ietf.org/html/rfc6265#section-3).
+"""
+
+function appendheader(m::Message, header::Pair{String,String})
+    c = m.headers
+    k,v = header
+    if k == ""
+        c[end] = c[end][1] => string(c[end][2], v)
+    elseif k != "Set-Cookie" && length(c) > 0 && k == c[end][1]
+        c[end] = c[end][1] => string(c[end][2], ", ", v)
+    else
+        push!(m.headers, header)
+    end
+    return
+end
+
+
+"""
+    httpversion(::Message)
+
+e.g. `"HTTP/1.1"`
+"""
+
+httpversion(m::Message) = "HTTP/$(m.version.major).$(m.version.minor)"
+
+
+"""
+    writestartline(::IO, ::Message)
+
+e.g. `"GET /path HTTP/1.1\\r\\n"` or `"HTTP/1.1 200 OK\\r\\n"`
+"""
+
+function writestartline(io::IO, r::Request)
+    write(io, "$(r.method) $(r.uri) $(httpversion(r))\r\n")
+    return
+end
+
+function writestartline(io::IO, r::Response)
+    write(io, "$(httpversion(r)) $(r.status) $(statustext(r))\r\n")
+    return
+end
+
+
+"""
+    writeheaders(::IO, ::Message)
+
+Write `Message` start line and
+a line for each "name: value" pair and a trailing blank line.
+"""
+
+function writeheaders(io::IO, m::Message)
+    writestartline(io, m)                 # FIXME To avoid fragmentation, maybe
+    for (name, value) in m.headers        # buffer header before sending to `io`
+        write(io, "$name: $value\r\n")
+    end
+    write(io, "\r\n")
+    return
+end
+
+
+"""
+    write(::IO, ::Message)
+
+Write start line, headers and body of HTTP Message.
+"""
+
+function Base.write(io::IO, m::Message)
+    writeheaders(io, m)
+    write(io, m.body)
+    return
+end
+
+
+function Base.String(m::Message)
+    io = IOBuffer()
+    write(io, m)
+    String(take!(io))
+end
+
+
+"""
+    readstartline!(::Parsers.Message, ::Message)
+
+Read the start-line metadata from Parser into a `::Message` struct.
+"""
+
+function readstartline!(m::Parsers.Message, r::Response)
+    r.version = VersionNumber(m.major, m.minor)
+    r.status = m.status
+    return
+end
+
+function readstartline!(m::Parsers.Message, r::Request)
+    r.version = VersionNumber(m.major, m.minor)
+    r.method = m.method
+    r.uri = m.url
+    return
+end
+
+
+"""
+    readheaders(::IO, ::Parser, ::Message)
```

Comment:
why not `readheaders!` to match `readstartline!`?

Reply:
that sounds like a fair argument